File: tests/DbDataSync.Web.Tests/mapping-load-waiter.ts
import type { Page } from '@playwright/test'

/**
 * Phase 134: a mapping whose reader captures its own position (Change Tracking, CDC, TriggerAudit)
 * never loads on its own first Primary pass — that pass only captures a position and requests a Bulk
 * Load, which runs concurrently. A run list containing a terminal `Succeeded` Primary run therefore
 * does not mean the data has landed; a test that reads real target rows right after seeing `Succeeded`
 * races that Bulk Load.
 *
 * This is the TypeScript side of `MappingLoadWaiter.WaitForLoadToCompleteAsync`
 * (tests/DbDataSync.Api.Tests/MappingLoadWaiter.cs, phase 141) — same idea, same endpoint: poll the
 * mapping's own read-state until its hold clears, since `SchedulerService.FilterHeld` is the one place
 * that ever waits for this and it does the identical check before handing a real mapping a further pass.
 */
export async function waitForLoadToComplete(
  page: Page, replicationName: string, mappingName: string, timeoutMs = 30_000,
): Promise<void> {
  const deadline = Date.now() + timeoutMs
  while (Date.now() < deadline) {
    const response = await page.request.get(
      `/api/replications/${replicationName}/table-mappings/${mappingName}/read-state`)
    const state: { hold: string } = await response.json()
    if (state.hold !== 'Loading') return
    await new Promise((resolve) => setTimeout(resolve, 250))
  }

  throw new Error(`'${mappingName}' on '${replicationName}' was still Loading after ${timeoutMs}ms.`)
}

File: tests/DbDataSync.Api.Tests/MappingLoadWaiter.cs
using System.Net.Http.Json;
using System.Text.Json;
using DbDataSync.Api.Controllers;
using DbDataSync.Core.Config;

namespace DbDataSync.Api.Tests;

/// <summary>
/// Phase 134: a mapping whose reader captures its own position (Change Tracking, CDC, TriggerAudit)
/// never loads on its own first Primary pass — that pass only captures a position and requests a Bulk
/// Load, which the real worker process these Integration-tagged tests spawn (see
/// <see cref="TestApiFactory"/>'s own doc) drains concurrently. Waiting for the *Primary* run's own row
/// to go terminal — what every test using this pattern already did before phase 134 — is not the same
/// as waiting for that Bulk Load to actually finish; a test that reads real target rows, or triggers a
/// further pass, immediately afterward races it.
/// <para>
/// This is the same thing <c>SchedulerService.FilterHeld</c> already waits for before ever handing a
/// real mapping a further Primary pass (<see cref="ReadHold.Loading"/> — phase 134's own retrospective
/// is explicit that this is the only enforcement point, by design): poll the mapping's own read-state
/// endpoint until its hold clears.
/// </para>
/// </summary>
internal static class MappingLoadWaiter
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
        Converters = { new System.Text.Json.Serialization.JsonStringEnumConverter() },
    };

    public static async Task WaitForLoadToCompleteAsync(
        this HttpClient client, string replicationName, string mappingName, TimeSpan? timeout = null)
    {
        var deadline = DateTimeOffset.UtcNow.Add(timeout ?? TimeSpan.FromSeconds(30));
        while (DateTimeOffset.UtcNow < deadline)
        {
            var response = await client.GetAsync(
                $"/api/replications/{replicationName}/table-mappings/{mappingName}/read-state");
            response.EnsureSuccessStatusCode();
            var state = await response.Content.ReadFromJsonAsync<MappingReadStateDto>(JsonOptions);
            if (state!.Hold != ReadHold.Loading)
                return;
            await Task.Delay(250);
        }

        throw new TimeoutException(
            $"'{mappingName}' on '{replicationName}' was still Loading after " +
            $"{(timeout ?? TimeSpan.FromSeconds(30)).TotalSeconds}s.");
    }
}
